wasm/execution/runtime.py

```diff
@@ -0,0 +1,251 @@
+from typing import (
+    Dict,
+    Iterator,
+    Optional,
+    Tuple,
+    Union,
+    cast,
+)
+
+from wasm.datatypes import (
+    DataSegment,
+    ElementSegment,
+    FunctionAddress,
+    GlobalAddress,
+    Import,
+    MemoryAddress,
+    Module,
+    ModuleInstance,
+    Store,
+    TableAddress,
+)
+from wasm.exceptions import (
+    InvalidModule,
+    Unlinkable,
+    ValidationError,
+)
+from wasm.typing import (
+    TValue,
+    UInt32,
+)
+from wasm.validation import (
+    validate_external_type_match,
+    validate_module,
+)
+
+from .configuration import (
+    Configuration,
+)
+from .instructions import (
+    InstructionSequence,
+)
+from .stack import (
+    Frame,
+)
+
+TAddress = Union[FunctionAddress, TableAddress, MemoryAddress, GlobalAddress]
+
+
+def _get_import_addresses(runtime: 'Runtime',
+                          imports: Tuple[Import, ...]) -> Iterator[TAddress]:
+    for import_ in imports:
+        if not runtime.has_module(import_.module):
+            raise Unlinkable(f"Runtime has no known module named '{import_.module}'")
+        module = runtime.get_module(import_.module)
+        for export in module.exports:
+            if export.name == import_.name:
+                yield export.value
+                break
+        else:
+            raise Unlinkable(
+                f"No export found with name '{import_.name}'"
+            )
+
+
+def _initialize_globals(store: Store,
+                        module: Module,
+                        globals_addresses: Tuple[GlobalAddress, ...],
+                        ) -> Iterator[UInt32]:
+        module_instances = ModuleInstance(
+            types=(),
+            func_addrs=(),
+            memory_addrs=(),
+            table_addrs=(),
+            global_addrs=globals_addresses,
+            exports=(),
+        )
+
+        for global_ in module.globals:
+            config = Configuration(store=store)
+            frame = Frame(
+                module=module_instances,
+                locals=[],
+                instructions=InstructionSequence(global_.init),
+                arity=1,
+            )
+            config.push_frame(frame)
+            result = config.execute()
+            if len(result) != 1:
+                raise Exception("Invariant: globals initialization returned empty result")
```

Comment:
Didn't you add a `get_single_or_crash()` kind of thing somewhere?

Reply:
Yes, but it's a bit more awkward to implement in a generic way since the error messages need to be different in each case.  Similar to the other one, this is a runtime check which we should never be able to hit since this would be caught during validation.